rpass.1
'\" t
.\"     Title: rpass
.\"    Author: Yuri D. Lenskiy <yuri.lenskiy@gmail.com>
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 12/27/2010
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "RPASS" "1" "12/27/2010" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
rpass \- secure (i\&.e\&. encrypted), simple, command\-line password manager
.SH "SYNOPSIS"
.sp
\fBrpass\fR \-a [\fINEW ACCOUNT NAMES\fR \&...]
.sp
\fBrpass\fR \-d \fIACCOUNT NAMES\fR
.sp
\fBrpass\fR [\fIOPTIONS\fR] \fIPATTERNS\fR \&...
.sp
\fBrp\fR
.sp
\fBru\fR
.SH "DESCRIPTION"
.sp
\fBrpass\fR is a module and related executable that are meant to make secure password management through the commandline and almost all programs that use passwords accessible, convenient, and easy\&.
.SH "DEPENDENCIES"
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
The executable
\fBgpg\fR
must be in your $PATH\&. This is the ONLY required dependency\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
In order for copying to the X\-clipboard to work,
\fBxclip\fR
must be in the user\(cqs $PATH\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}

\fBgpg\-agent\fR
for tool to be used to full potential\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}

\fBdmenu\fR
for gui tools (\fBrp\fR
and
\fBru\fR)
.RE
.SH "SETUP"
.sp
If installing from source:
.sp
.if n \{\
.RS 4
.\}
.nf
# python setup\&.py install
.fi
.if n \{\
.RE
.\}
.sp
Run rpass once to set up \fBgpg\fR (generate keys) if this has not been done already\&.
.SH "USAGE"
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.sp
See rpass \-h for a complete summary of command\-line options\&.
.sp .5v
.RE
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBWarning\fR
.ps -1
.br
.sp
It is highly recommended to have \fBxclip\fR installed so that passwords never have to be written to stdout\&.
.sp .5v
.RE
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.sp
Passwords and usernames are never printed unless stated explicitly (see rpass \-h)
.sp .5v
.RE
.SS "CONFIGURATION"
.sp
Upon installation a heavily commented configuration file is placed in the home directory at ~/\&.rpass\&.conf\&. If you delete or change this configuration file and want to see the comments/documentation again, there is one located in $PREFIX/share/rpass/rpass\&.example\&.conf\&.
.sp
In this configuration file you can set up most of the relevant defaults to \fBrpass\fR\&.
.SS "VIEWING"
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.sp
These are the defaults, you can change them easily in the configuration file
.sp .5v
.RE
.sp
The command \fBrpass\fR on its own will list account names:
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass
Account Name 1
Account Name 2
Account Name 3
\&.\&.\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.sp
No password is copied if \fBrpass\fR is run without an argument\&.
.sp .5v
.RE
.sp
The command \fBrpass\fR followed by any text will print any accounts that contain the text in the account name (search is case\-insensitive with regular expressions allowed) with all optional fields shown, and copy the password of the first matching account to the clipboard:
.sp
Simple query:
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass 1
Account Name 1
\&.\&.\&.
.fi
.if n \{\
.RE
.\}
.sp
Regular expressions:
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass "name (1|3)$"
Account Name 1
Account Name 3
    optional_field1: optional_value1
\&.\&.\&.
.fi
.if n \{\
.RE
.\}
.SS "SHOWING INFORMATION"
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.sp
Any fields can be shown in the default view (the most potentially useful being username) by editing the configuration file (~/\&.rpass\&.conf)\&.
.sp .5v
.RE
.sp
Username:
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-\-users [Account searches]
.fi
.if n \{\
.RE
.\}
.sp
or
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-u [Account searches]
.fi
.if n \{\
.RE
.\}
.sp
Password:
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-\-passwords [Account searches]
.fi
.if n \{\
.RE
.\}
.sp
or
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-p [Account searches]
.fi
.if n \{\
.RE
.\}
.sp
Custom fields:
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-\-keys \*(Aqcustom1,custom2\*(Aq [Account searches]
.fi
.if n \{\
.RE
.\}
.sp
You can combine these in any way:
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-puk \*(Aqcustom1\*(Aq [Account searches]
.fi
.if n \{\
.RE
.\}
.SS "PIPING"
.sp
rpass attempts intelligent piping support by removing color formatting when it is being piped\&.
.sp
An example is the included utility \fBrp\fR\&. It used dmenu to select an rpass entry\&. It\(cqs code is the following:
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass "$(rpass|dmenu \-i)"
.fi
.if n \{\
.RE
.\}
.SS "EDITING"
.sp
The command \fBrpass \-a\fR brings up an interface to add accounts and information\&. You can optionally list the account name(s) on the commandline\&.
.sp
Creates one account (enter name when prompted):
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-a
.fi
.if n \{\
.RE
.\}
.sp
Creates one account named "Secure Account1" (will be prompted for everything but account name):
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-a "Secure Account1"
.fi
.if n \{\
.RE
.\}
.sp
Creates three accounts (will be prompted for everything but account names):
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-a "Secure Account2" account3 "Account Name 4"
.fi
.if n \{\
.RE
.\}
.sp
The command \fBrpass \-d\fR followed by exact account names deletes those accounts:
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-d "Secure Account2" account3
.fi
.if n \{\
.RE
.\}
.SS "PLUGIN"
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBWarning\fR
.ps -1
.br
.sp
For rpass to be effective as a plugin, you need \fBgpg\-agent\fR
.sp .5v
.RE
.sp
For any python program or program with support for python plugins, "import rpass" will allow you to use all the functions of the main program\&.
.sp
For non\-python programs and scripts the \-\-batch and \-\-keys options allow the user to select information to print for any account\&.
.sp
The \fB\-\-batch\fR option forces rpass to do exact matching on account names and eliminates all formatting and field labels\&. The \fB\-\-keys\fR option allows the user to specify custom keys to print\&.
.SS "BATCH EXAMPLES"
.sp
Get the unformatted password from account named "AC ONE":
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-\-batch \-\-password "AC ONE"
secretpass123
.fi
.if n \{\
.RE
.\}
.sp
Get the unformatted username and password from account named "AC ONE":
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-\-batch \-\-users \-\-password "AC ONE"
username
secretpass123
.fi
.if n \{\
.RE
.\}
.sp
or
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-\-bpu "AC ONE"
username
secretpass 123
.fi
.if n \{\
.RE
.\}
.sp
Get the unformatted username from account named "AC ONE":
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-\-batch \-\-keys \*(Aquser\*(Aq "AC ONE"
acusername
.fi
.if n \{\
.RE
.\}
.sp
Get the unformatted custom field "custom1" and the username from account named "AC ONE":
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-\-batch \-\-users \-\-keys \*(Aqcustom1\*(Aq "AC ONE"
custom1value
acusername
.fi
.if n \{\
.RE
.\}
.sp
or
.sp
.if n \{\
.RS 4
.\}
.nf
# rpass \-buk \*(Aqcustom1\*(Aq "AC ONE"
custom1value
acusername
.fi
.if n \{\
.RE
.\}
.SH "SEE ALSO"
.sp
gpg\-agent(1), gpg(1)
.SH "GOALS"
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Secure
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Passwords are never written out in plaintext to harddrive under ANY conditions
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
With
\fBxclip\fR
installed, passwords never need to even be seen
.RE
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Simple user interface
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Automatic copying
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Intuitive, simple searching
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Powerful regular expression searching
.RE
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Transparent code
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Simple codebase allows bugs to surface quickly, adding to security
.RE
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Pluggable
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Importable python module for integration to other python programs
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Batch mode for executable
.RE
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Universal password solution
.RE
.SH "AUTHOR"
.PP
\fBYuri D\&. Lenskiy\fR <\&yuri\&.lenskiy@gmail\&.com\&>
.RS 4
Author.
.RE


rpass.py
class UnencryptedFile(IOError):
    """Exception to take care of unencrypted files."""
    pass

class ExistingEntry(ValueError):
    """Exception raised when attempting to overwrite an existing entry."""
    pass

class NonexistentEntry(ValueError):
    """Exception raised when attempting to modify nonexistent entry."""
    pass

class InvalidEncryptionKey(ValueError):
    """Exception raised when key is nonexistent or incorrect."""
    pass

def IsRunning(procname):
    import re
    from subprocess import Popen,PIPE

    ppatt = re.compile(r'^\s*\d+ \S*\s*\d\d:\d\d:\d\d (.*)$')
    proc = Popen(['ps', '-e'], stdout = PIPE, stderr = PIPE)

    plist = [re.match(ppatt, p.strip()).groups()[0] for p in str(proc.communicate()[0], encoding='utf-8').split('\n')[1:-1]]

    if procname in plist: return True
    else: return False

class rpass:
    """ Class to encapsulate an rpass user """

    def __init__(self, conf_file = '~/.rpass.conf', passfile = None, executable = False):
        """Initialized rpass from an optional configuration file and arguments.
        
        Set executable to 'true' if instance is not being used for a plugin."""

        from os.path import isfile,expanduser

        if not(self.HasGPGKey()):
            self.CreateGPGKey()

        self.executable = executable
        self.conf_file = conf_file
        (self.options, self.copyerror) = self.ReadConfigFile(filename = self.conf_file)

        if passfile != None: self.options['passfile'] = passfile
        if ('passfile' not in self.options) or not(self.options['passfile']): self.options['passfile'] = '~/.passwords.gpg'

        self.passfile = expanduser(self.options['passfile'])

        if not(isfile(self.passfile)): self.first = True
        else: self.first = False

        self.entries = None
        if not(self.first):
            self.entries = self.ParsePassFile(self.passfile)

    def CreateGPGKey(self):
        """Creates a gpg key for the user."""

        from subprocess import call
        print("Need to first create gpg key pair.")
        print("Choose a secure passphrase -- this is going to be your 'master' password.")
        print("Rerun program after key creation.\n")
        print("WARNING: DO NOT pick a sign-only key type.\n")
        print("-------------------------------------------------------------------------")
        input("Press [ENTER] when ready.")
        call('gpg --gen-key'.split(' '))

    def HasGPGKey(self):
        """Checks if the user has a gpg key pair."""

        from os import stat
        from os.path import expanduser,isfile,join

        secring = join(expanduser('~'), '.gnupg', 'secring.gpg')
        if not(isfile(secring)) or stat(secring).st_size == 0: return False
        else: return True

    def CopyDefaultConf(self, location, conf_locs):
        """Attempts to copy the default configuration to the specified location."""

        from os.path import expanduser,isfile
        fexample = None
        for p in conf_locs:
            if isfile(p): 
                fexample = p
                break

        if not(fexample):
            raise IOError("Could not find default configuration file in any of the following locations: {0}".format(','.join(conf_locs)))
        else:
            from shutil import copyfile
            copyfile(fexample, expanduser(location))

    def ReadConfigFile(self, filename, info = {}):
        """When passed a filename, reads the config file for rpass information.
        
        Returns a dictionary of useful values."""
        import configparser
        from os.path import expanduser,isfile

        if not(isfile(expanduser(filename))):

            # Try to copy the default configuration
            from sys import path

            paths = ["{0}/../share/rpass/", "/usr/share/rpass/", "/usr/local/share/rpass/", "{0}/"]
            paths = [p.format(path[0]) + 'rpass.example.conf' for p in paths]

            try:
                self.CopyDefaultConf(expanduser(filename), paths)
            except IOError as e:
                info['fields'] = []
                return (info, str(e))

        config = configparser.ConfigParser()
        config.read(expanduser(filename))

        if config.has_option('display', 'fields'): 
            info['fields'] = [field.strip() for field in 
                    config.get('display', 'fields').split(',') if field.strip()]
            if len([f for f in info['fields'] if f.strip()]) == 0: info['fields'] = [None]
        if config.has_option('display', 'color'):
            if config.get('display', 'color').lower()[0] == 'y': info['color'] = True
            else: info['color'] = False

        if config.has_option('general', 'passfile'):
            from os.path import expanduser
            info['passfile'] = expanduser(config.get('general', 'passfile'))

        return (info, None)

    def DecryptPassFile(self, passfile):
        from os.path import isfile,expanduser
        if not isfile(passfile): raise IOError("Password file not found.")

        from subprocess import Popen,PIPE
        from os import environ

        gpg_info_name = expanduser('~/.gpg-agent-info')
        has_gpg_info = isfile(gpg_info_name)
        proclst = ['gpg', '--quiet', '--output', '-', '--decrypt', passfile]

        if not(self.executable):
            if not(IsRunning('gpg-agent')):
                raise RuntimeError("No gpg-agent running when rpass is being used as a plugin.")
            if not(has_gpg_info):
                raise RuntimeError("gpg-agent invoked without --write-env-file option when rpass is being used as a plugin.")

        if IsRunning('gpg-agent'): 
            proclst.insert(1, '--no-tty')
            proclst.insert(1, '--use-agent')
            if has_gpg_info:
                with open(gpg_info_name) as GPG_FILE:
                    tmp = GPG_FILE.readlines()
                    environ.update(dict([tuple(t.strip().split('=')) for t in tmp]))

        proc = Popen(proclst, stdout = PIPE, stderr = PIPE, env = environ)
     
        retstr, errstr = tuple(str(s, encoding = "utf-8") for s in proc.communicate())
        if errstr.find("gpg: no valid OpenPGP data found.") != -1: raise UnencryptedFile
        elif (errstr.find("gpg: decryption failed: secret key not available") != -1) or (errstr.find("gpg: decrypt_message failed: eof") != -1): raise InvalidEncryptionKey

        return retstr.strip()

    def EncryptPassFile(self, passfile, contents):
        from subprocess import Popen,PIPE
        textproc = Popen(['echo', contents], shell=False, stdout=PIPE)
        encproc = Popen(['gpg', '--default-recipient-self', '--yes', '--output', passfile, '--encrypt'], shell=False, stdin=textproc.stdout).wait()

    def Write(self):
        self.EncryptPassFile(self.passfile, self.CreatePassFile(self.entries))

    def ParsePassFile(self, passfile, contents = None):
        if contents == None: contents = self.DecryptPassFile(passfile = passfile)

        import re
        parray = [s.strip() for s in contents.split('\n') if not(re.match(r'^\s*$', s))]
        pdict = {}
        ckey = ""
        accountpatt = re.compile(r'^\[([^]]+)\]$')
        fieldpatt = re.compile(r'^(.*?)\s*=\s*(.*)$')
        for i in parray:
            if accountpatt.match(i) != None:
                ckey = accountpatt.match(i).group(1)
                pdict[ckey] = {}
            else:
                match = fieldpatt.match(i)
                pdict[ckey][match.group(1)] = match.group(2)

        return pdict

    def GetAccountInfo(self, account, strict = False):
        if self.entries == None:
            return {}

        import re
        accountpatt = ''
        if strict:
            accountpatt = re.compile("^{0}$".format(account))
        else:
            accountpatt = re.compile(account, re.I)
        accountdict = {}

        for ac in self.entries.keys():
            if accountpatt.search(ac): accountdict[ac] = self.entries[ac]

        return accountdict

    def CopyPass(self, acinfo = None, account = '.'):
        if acinfo == None: return False
        from subprocess import Popen,PIPE

        for ac in sorted(acinfo.keys()):
            if 'pass' in acinfo[ac]:
                echoproc = Popen(['echo', acinfo[ac]['pass']], shell=False, stdout=PIPE)
                copyproc = Popen(['xclip'], shell=False, stdin=echoproc.stdout)
                return True
        return False

    def CreatePassFile(self, pinfo):
        contents = ""
        ftemplate = "    {0} = {1}\n"
        sorted_accounts = sorted(pinfo.keys())
        for key in sorted_accounts:
            contents += "[{0}]\n".format(key)
            if "user" in pinfo[key]: contents += ftemplate.format("user", pinfo[key]["user"])
            if "pass" in pinfo[key]: contents += ftemplate.format("pass", pinfo[key]["pass"])
            for (field, value) in pinfo[key].items():
                if field not in ["user", "pass"]: contents += ftemplate.format(field, value)
            contents += "\n"
        return contents.strip()

    def DeleteEntry(self, entry):
        if entry in self.entries: 
            del(self.entries[entry])
        else:
            raise NonexistentEntry

    def AddEntry(self, name, entry):
        """Adds an entry to the account."""

        if type(self.entries) == dict:
            if name in self.entries: raise ExistingEntry
        else:
            self.entries = {}

        self.entries[name] = entry
